In [1]:
%conda install matplotlib -y

import pandas as pd
import numpy as np
import statsmodels.formula.api as sm

from sklearn import datasets
from sklearn.linear_model import LinearRegression

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 23.3.1
  latest version: 23.10.0

Please update conda by running

    $ conda update -n base -c defaults conda

Or to minimize the number of packages updated during conda update use

     conda install conda=23.10.0



# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.


# Multiple Linear Regression - practice

## Multiple Linear Regression 

The following are some of the data regarding diabetic patients. Propose a model that describes diabetes levels using all of your age, BMI, and blood pressure (bp), and test properly.

In [2]:
dia = datasets.load_diabetes()

dia = pd.DataFrame(dia['data'], index=dia['target'], columns=dia['feature_names'])
dia = dia.reset_index().rename(columns={'index': 'dia'})
display(dia)

dia       age       sex       bmi        bp        s1        s2  \
0    151.0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821   
1     75.0 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163   
2    141.0  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194   
3    206.0 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991   
4    135.0  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596   
..     ...       ...       ...       ...       ...       ...       ...   
437  178.0  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566   
438  104.0 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165   
439  132.0  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840   
440  220.0 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283   
441   57.0 -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809   

           s3        s4        s5        s6  
0   -0.043401 -0.002592  0.019907 -0.017646  
1    0.074412 -0.039493 -0.068332 -0.092204  
2   -0.032356 -0.002592  0.002861 -0.025930  
3   -0.036038  0.034309  0.022688 -0.009362  
4    0.008142 -0.002592 -0.031988 -0.046641  
..        ...       ...       ...       ...  
437 -0.028674 -0.002592  0.031193  0.007207  
438 -0.028674  0.034309 -0.018114  0.044485  
439 -0.024993 -0.011080 -0.046883  0.015491  
440 -0.028674  0.026560  0.044529 -0.025930  
441  0.173816 -0.039493 -0.004222  0.003064  

[442 rows x 11 columns]

After extracting only the independent variables and diabetes levels to be used, the model calculated using the least squares method is as follows.

In [5]:
dia = dia[['age', 'bmi', 'bp', 'dia']]
display(dia)

lr = LinearRegression()
lr.fit(dia[['age', 'bmi', 'bp']].values, dia['dia'].to_numpy().reshape(-1, 1))

print(f"Model: y = {lr.intercept_[0]} + {lr.coef_[0][0]} (age) + {lr.coef_[0][1]} (bmi) + {lr.coef_[0][2]} (bp) + e")

age       bmi        bp    dia
0    0.038076  0.061696  0.021872  151.0
1   -0.001882 -0.051474 -0.026328   75.0
2    0.085299  0.044451 -0.005670  141.0
3   -0.089063 -0.011595 -0.036656  206.0
4    0.005383 -0.036385  0.021872  135.0
..        ...       ...       ...    ...
437  0.041708  0.019662  0.059744  178.0
438 -0.005515 -0.015906 -0.067642  104.0
439  0.041708 -0.015906  0.017293  132.0
440 -0.045472  0.039062  0.001215  220.0
441 -0.045472 -0.073030 -0.081413   57.0

[442 rows x 4 columns]

Model: y = 152.13348416289614 + 25.990394294685952 (age) + 788.7826891866822 (bmi) + 394.1270922265727 (bp) + e


Null hypothesis: there is no linear relationship between the model and the dependent variable.  
Alternative Hypothesis: There is a linear relationship between at least one independent variable and the dependent variable.

In [4]:
results = sm.ols(formula='dia ~ age + bmi + bp', data = dia).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    dia   R-squared:                       0.396
Model:                            OLS   Adj. R-squared:                  0.392
Method:                 Least Squares   F-statistic:                     95.81
Date:                Sat, 18 Nov 2023   Prob (F-statistic):           1.09e-47
Time:                        18:47:11   Log-Likelihood:                -2435.7
No. Observations:                 442   AIC:                             4879.
Df Residuals:                     438   BIC:                             4896.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    152.1335      2.859     53.211      0.000     146.514     157.753
age           25.9904     63.923      0.407      0.685     -99.643     151.623
bmi          788.7827     65.562     12.031      0.000     659.927     917.638
bp           394.1271     68.392      5.763      0.000     259.710     528.544
==============================================================================
Omnibus:                       12.586   Durbin-Watson:                   1.928
Prob(Omnibus):                  0.002   Jarque-Bera (JB):                7.934
Skew:                           0.175   Prob(JB):                       0.0189
Kurtosis:                       2.444   Cond. No.                         28.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Interpretation  
First, the coefficient of determination is 0.396, and the adjusted coefficient of determination is 0.392, so the model hardly explains the data.  
On the other hand, since the p-value of the F test is close to zero, there is sufficient evidence to reject the null hypothesis at the 99.99% confidence level.  
=> A linear relationship exists between at least one independent variable and the dependent variable.

- Age: Since the p-value is 0.41, there is insufficient evidence to reject the null hypothesis.  
=> There is no linear relationship between age and diabetes.

- BMI, bp: p-value is 0.00 so there is sufficient evidence to reject the null hypothesis at a 99.99% confidence level.  
=> There is a linear relationship between BMI and diabetes. In addition, a linear relationship exists between bp and diabetes.